File: app/api/db/mongo/mongo.py
import pprint
from pymongo import AsyncMongoClient
from gridfs.asynchronous import AsyncGridFS

from .utils import TS, SEQ
from .data import DocData, DocDataUpdate
from .lib_files import LibFiles

LIB_DOC_CLASS = 'LIB:DOC'

class Lib:

    def __init__(self, connection_str):
        self.data = []
        self.client = AsyncMongoClient(connection_str)
        self.db = self.client.ph3
        self.fs = AsyncGridFS(self.db, collection="docs")
        self.docs = self.db["docs"]
        self.files = self.db["docs.files"]
        self.lib_files = LibFiles(fs=self.fs, files=self.files)

    async def close(self):
        await self.client.close()

    async def get_children(self, id: str | None):
        if id == "": 
            id = None
        cursor = self.docs.find({
            '_meta.class': LIB_DOC_CLASS,
            '_meta.parent': id
        }).sort('_meta.ctime', -1)
        children = await cursor.to_list(length=None)
        return [DocData(**c) for c in children]

    # async def get_path(self, id: str | None):
    async def get_path(self, id: str):
        # print("get_path 001: ", id)
        rez = []
        current_id = id
        while current_id:
            d = await self.get_data(current_id)
            if d and d.meta and d.meta.parent:
                rez.insert(0, d)
                current_id = d.meta.parent
            else:
                current_id = None
        return rez

    # async def get_data(self, id: str | None):
    async def get_data(self, id: str):
        # pprint.pprint(">>>>>>>>>>")
        pprint.pprint("get_data:")
        # pprint.pprint("id:")
        # pprint.pprint(id)
        if id is not None:
            doc = await self.docs.find_one({"_id": id})
            pprint.pprint("doc:")
            pprint.pprint(doc)
            rez = DocData(**doc) if doc else None
            pprint.pprint("rez:")
            pprint.pprint(rez)
        else:
            rez = None
        # pprint.pprint("<<<<<<<<<<")
        return rez

    # async def set_data(self, id: str | None, data: dict):
    # async def set_data(self, id: str | None, data: DocDataUpdate):
    async def set_data(self, id: str, data: DocDataUpdate):
        # print(">>> set_data", id, data)
        # data = data.model_dump(exclude_unset=True)
        newvalues = {"$set": data.model_dump()}
        # newvalues = {"$set": {
        #     'info': data.get('info'),
        #     'authors': data.get('authors'),
        # }}
        await self.docs.update_one({"_id": id}, newvalues)
        return {'saved': True}

    async def new_doc(self, parent, info, args=None):
        args = args or {}

        _id = SEQ()
        ts = TS()

        meta = {
            "class": LIB_DOC_CLASS,
            "parent": parent,
            "ctime": ts,
            "mtime": ts
        }

        if args.get("origin"):
            meta["origin"] = args["origin"]

        authors = args.get("authors", [])

        document = {
            "_id": _id,
            "_meta": meta,
            "info": info,
            "authors": authors
        }

        await self.docs.insert_one(document)
        return _id
    

File: lib_go/app/data/mongo/mongo.go
package mongo

import (
	"io"
	"context"
	"time"
	// "log"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/gridfs"
	"go.mongodb.org/mongo-driver/mongo/options"

	. "lib/data/types"
)

type MongoDB struct {
	Client   *mongo.Client
	DB       *mongo.Database
	Docs     *mongo.Collection
	Files    *mongo.Collection
	GridFS   *gridfs.Bucket
}

func ConnectMongoDB(uri string) (*MongoDB, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	client, err := mongo.Connect(ctx, options.Client().ApplyURI(uri))
	if err != nil {
		return nil, err
	}

	dbName := "ph3"

	db := client.Database(dbName)
	bucket, err := gridfs.NewBucket(db, options.GridFSBucket().SetName("docs"))
	if err != nil {
		return nil, err
	}

	return &MongoDB{
		Client: client,
		DB:     db,
		Docs:   db.Collection("docs"),
		Files:  db.Collection("docs.files"),
		GridFS: bucket,
	}, nil
}

func (db *MongoDB) FindFullDoc(ctx context.Context, id string) (*Document, error) {
	var (
		doc Document
		fileInfo FileInfo
	)

	err := db.Docs.FindOne(ctx, bson.M{"_id": id}).Decode(&doc)
	if err != nil { return nil, err }

	err = db.Files.FindOne(ctx, bson.M{"_meta.parent": id}).Decode(&fileInfo)
	if err == nil { doc.FileInfo = fileInfo }

	doc.Children, _ = db.FindDocs(ctx, bson.M{"_meta.parent": id}, bson.D{{"_meta.ctime", -1}})
	doc.Breadcrumbs, _ = db.GetDocBreadcrumbs(ctx, id)

	return &doc, nil
}


func (db *MongoDB) FindDocs(ctx context.Context, filter bson.M, sort bson.D) ([]Document, error) {
	opts := options.Find().SetSort(sort)
	cursor, err := db.Docs.Find(ctx, filter, opts)
	if err != nil {
		return nil, err
	}

	var results []Document
	if err = cursor.All(ctx, &results); err != nil {
		return nil, err
	}
	return results, nil
}

func (db *MongoDB) GetDocBreadcrumbs(ctx context.Context, id string) ([]Breadcrumb, error) {
	var (
		doc Document
		result []Breadcrumb
		currentId = id
	)
	for {
		err := db.Docs.FindOne(ctx, bson.M{"_id": currentId}).Decode(&doc)
		if err != nil { return nil, err }

		current := Breadcrumb{ID: currentId, Title: doc.Info.Title}
		result = append([]Breadcrumb{current}, result...)

		currentId = doc.Meta.Parent
		if currentId=="" { break }
	}
	return result, nil
}

func (db *MongoDB) DownloadFile(ctx context.Context, id string, w io.Writer) error {
	stream, err := db.GridFS.OpenDownloadStream(id)
	if err != nil {
		return err
	}
	defer stream.Close()

	_, err = io.Copy(w, stream)
	return err
}
